src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -266,7 +266,13 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $fragment = isset($extra['_fragment']) ? $extra['_fragment'] : '';
         unset($extra['_fragment']);
 
-        if ($extra && $query = http_build_query($extra, '', '&')) {
+        if (defined('PHP_QUERY_RFC3986')) {
```

Comment:
5.4, the reason I've done it this way is to ease merging of upstream changes on 2.x branch